### PR TITLE
feat: support EuphonyRxPanel component (#237)

### PR DESCRIPTION
**Pull request by haan823**
I made EuphonyRxPanel component for euphony user.

They will use euphony easily compared to before.

`EuphonyRxPanel` has two components.
- listen button
- result text field

## Initial State
image

## Listening State
image

## Success Result
image

## Fail Result
image

## Build & Test Result
image

### CODE REVIEW

File: euphony/src/androidTest/java/co/euphony/ui/EuphonyRxPanelTest.kt [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+class EuphonyRxPanelTest {
+
+    @get:Rule
+    val composeTestRule = createComposeRule()
+
+    @Mock
+    private lateinit var viewModel: EuphonyRxPanelViewModel
+
+    private val isListening = MutableStateFlow(false)
+    private val isListenStarted = MutableStateFlow(false)
```
Comment (yeonns) [inline euphony/src/androidTest/java/co/euphony/ui/EuphonyRxPanelTest.kt:35]: I think `isListenStarted` is not needed, so we can eliminate this variable :smile:

File: euphony/src/main/java/co/euphony/ui/EuphonyRxPanel.kt [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+@Composable
+fun EuphonyRxPanel() {
+    val viewModel = EuphonyRxPanelViewModel(EuRxManager())
+    EuphonyRxPanelImpl(viewModel = viewModel)
+}
+
+@Composable
+fun EuphonyRxPanelImpl(
+    viewModel: EuphonyRxPanelViewModel
+) {
```
Comment (yeonns) [inline euphony/src/main/java/co/euphony/ui/EuphonyRxPanel.kt:27]: I want user to use only `EuphonyRxPanel`, not `EuphonyRxPanelmpl`, so how about defining as `internal fun`?

Comment (yeonns) [inline euphony/src/main/java/co/euphony/ui/EuphonyRxPanel.kt:27]: I want user to use only `EuphonyRxPanel`, not `EuphonyRxPanelmpl`, so how about defining as `internal fun`?

File: euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+class EuphonyRxPanelViewModel(
+    private val rxManager: EuRxManager
+) : ViewModel() {
+
+    private val _isListening = MutableStateFlow(false)
+    val isListening: StateFlow<Boolean> = _isListening
+
+    private val _isListenStarted = MutableStateFlow(false)
+    val isListenStarted: StateFlow<Boolean> = _isListenStarted
```
Comment (yeonns) [inline euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt:20]: I think `isListenStarted` is not needed, so we can eliminate this variable :smile:

File: euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    private fun listen() {
+        rxManager.listen()
+        viewModelScope.launch {
+            while (limitTime > 0) {
+                delay(1000)
+                limitTime--
+            }
+            stop()
+        }
+    }
```
Comment (yeonns) [inline euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt:51]: `EuphonyRxPanel` is always waiting 10 seconds now. This is something to consider. :eyes:

**Review by yeonns — COMMENTED** [review]
@haan823
Thank you for implementing EuphonyRxPanel! Your code looks great :+1: 
Also, your test code is perfect :smile: 
I commented few things to be changed. Please check my comments :)

File: euphony/src/main/java/co/euphony/ui/EuphonyRxPanel.kt (hunk 2 above)
Comment (haan823) [reply euphony/src/main/java/co/euphony/ui/EuphonyRxPanel.kt:27]: Oh, i'll change the function definition 👍

File: euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt (hunk 4 above)
Comment (haan823) [reply euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt:51]: Oh, I missed it! What about making `input field` to user can define limit time?

File: euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt (hunk 3 above)
Comment (haan823) [reply euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt:20]: I think `isListenStarted` is needed to define textBackgroundColor in EuphonyRxPanel.

I intended to maintain background color to `LightSkyBlue` if user don't click `listen button` even once.

But, after user click the button more than once, background color must be one of 'LightGreen' or 'LightRed'.

Am I missing something? or could you recommend me better way?

Comment (yeonns) [reply euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt:20]: @haan823 Oh, I now checked your code where `isListenStarted` is used! :pray: 
Now, I understand why you declared `isListenStarted`.
Another way can be utilizing `limitTime`.
If `limitTime is <= 0`, it has result, so we can write code like below:
```kotlin
    val textBackgroundColor =
        if (!isListening && viewModel.getTime() <= 0) {
            if (viewModel.isSuccess()) {
                LightGreen
            } else {
                LightRed
            }
        } else {
            LightSkyBlue
        }
```

File: euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    private val _rxCode = MutableStateFlow("")
+    val rxCode: StateFlow<String> = _rxCode
+
+    private var limitTime = 10
+
+    fun start() {
+        if (!_isListening.value) {
+            _rxCode.value = ""
+            _isListening.value = true
+            _isListenStarted.value = true
```
Comment (yeonns) [inline euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt:31]:
```suggestion
            _rxCode.value = ""
            _isListening.value = true
            _isListenStarted.value = true
            // need below code
            limitTime = 10
```

I found one more!
If we do not re-initialize `limitTime` like above, we cannot execute EuphonyRxPanel multiple times.

File: euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt (hunk 3 above)
Comment (haan823) [reply euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt:20]: Oh, I think it is better code! Thank you 😄

File: euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt (hunk 5 above)
Comment (haan823) [reply euphony/src/main/java/co/euphony/ui/viewmodel/EuphonyRxPanelViewModel.kt:31]: I removed isListenStarted value & add re-initialize limitTime code

**Review by yeonns — COMMENTED** [review]
@haan823 Looks good!
https://github.com/euphony-io/euphony/pull/256 has just been merged, so I think the last thing to change is the declaration of 'EuRxManager'.

**Comment by haan823** [comment]
@yeonns I change EuRxManager using singleton!

**Review by yeonns — APPROVED** [review]
@haan823 
It was a long review, so thank you for applying all feedbacks from me :smile: 
I think it's enough :+1: 
Your EuphonyRxPanel will help people to utilize Euphony easily.
Thank you so much!

**Comment by haan823** [comment]
@yeonns 
Thank you so much for all the help this week. It was really helpful feedback. 😄

**Review by designe — APPROVED** [review]

**Pull request merged**